service.go

```diff
@@ -0,0 +1,250 @@
+		return nil, errTokenNotSet
+	}
+	return s, s.setupServiceClient()
+}
+
+// EndpointOption receives the TCP endpoint of mesg-core.
+func EndpointOption(address string) Option {
+	return func(s *Service) {
+		s.endpoint = address
+	}
+}
+
+// TokenOption receives token which is the unique id of this service.
+func TokenOption(token string) Option {
+	return func(s *Service) {
+		s.token = token
+	}
+}
+
+// TimeoutOption receives d to use while dialing mesg-core and making requests.
+func TimeoutOption(d time.Duration) Option {
+	return func(s *Service) {
+		s.callTimeout = d
+	}
+}
+
+// LogOutputOption uses out as a log destination.
+func LogOutputOption(out io.Writer) Option {
+	return func(s *Service) {
+		s.logOutput = out
+	}
+}
+
+// DialOption used to mock socket communication for unit testing.
+func DialOption(dialer Dialer) Option {
+	return func(s *Service) {
+		s.dialOptions = append(s.dialOptions, grpc.WithDialer(newGRPCDialer(dialer).Dial))
+	}
+}
+
+func (s *Service) setupServiceClient() error {
+	var err error
+	ctx, cancel := context.WithTimeout(context.Background(), s.callTimeout)
+	defer cancel()
+	s.conn, err = grpc.DialContext(ctx, s.endpoint, s.dialOptions...)
+	if err != nil {
+		return err
+	}
+	s.client = service.NewServiceClient(s.conn)
+	return nil
+}
+
+// Listen listens requests for given tasks. It's a blocking call.
+func (s *Service) Listen(task Taskable, tasks ...Taskable) error {
+	s.ml.Lock()
+	if s.isListening {
+		s.ml.Unlock()
+		return errAlreadyListening{}
+	}
+	s.isListening = true
+	s.ml.Unlock()
+	s.taskables = append(tasks, task)
+	if err := s.validateTasks(); err != nil {
+		return err
+	}
+	return s.listenTasks()
+}
+
+// validateTasks checks if the tasks handled exectly desribed in mesg.yaml.
+// TODO(ilgooz) use validation handlers of core server to do this?
+func (s *Service) validateTasks() error { return nil }
+
+func (s *Service) listenTasks() error {
+	var ctx context.Context
+	s.mc.Lock()
+	ctx, s.cancel = context.WithCancel(context.Background())
+	s.mc.Unlock()
+	stream, err := s.client.ListenTask(ctx, &service.ListenTaskRequest{
+		Token: s.token,
+	})
+	if err != nil {
+		return err
+	}
+	for {
+		s.gracefulWait.Add(1)
+		data, err := stream.Recv()
+		if err != nil {
+			s.gracefulWait.Done()
+			return err
+		}
+		go s.executeTask(data)
+	}
+}
+
+func (s *Service) getTaskableByName(key string) Taskable {
+	for _, taskable := range s.taskables {
+		if taskable.Name() == key {
+			return taskable
+		}
+	}
+	return nil
+}
+
+func (s *Service) executeTask(data *service.TaskData) {
+	defer s.gracefulWait.Done()
+	taskable := s.getTaskableByName(data.TaskKey)
+	if taskable == nil {
+		s.log.Println(errNonExistentTask{data.TaskKey})
+		return
+	}
+	execution := newExecution(s, data)
+	if err := execution.reply(taskable.Execute(execution)); err != nil {
+		s.log.Println(err)
+	}
+}
+
+// Emit emits a MESG event with given data for name.
+func (s *Service) Emit(event string, data Data) error {
```

Comment:
Can you add more context to the parameter: eg: EventKey, EventData.

Reply:
done